instrumentation/view-click/src/main/kotlin/io/opentelemetry/android/instrumentation/view/click/ViewClickEventGenerator.kt

```diff
@@ -0,0 +1,147 @@
+import java.lang.ref.WeakReference
+import java.util.LinkedList
+
+class ViewClickEventGenerator(
+    private val eventLogger: ExtendedLogger,
+) {
+    private var windowRef: WeakReference<Window>? = null
+
+    private val viewCoordinates = IntArray(2)
+
+    fun startTracking(window: Window) {
+        windowRef = WeakReference(window)
+        val currentCallback = window.callback
+        window.callback = WindowCallbackWrapper(currentCallback, this)
+    }
+
+    fun generateClick(motionEvent: MotionEvent?) {
+        windowRef?.get()?.let { window ->
+            if (motionEvent != null && motionEvent.actionMasked == MotionEvent.ACTION_UP) {
+                createEvent(APP_SCREEN_CLICK_EVENT_NAME)
+                    .setAttribute(yCoordinateAttr, motionEvent.y.toDouble())
+                    .setAttribute(xCoordinateAttr, motionEvent.x.toDouble())
+                    .emit()
+
+                findTargetForTap(window.decorView, motionEvent.x, motionEvent.y)?.let { view ->
+                    createEvent(VIEW_CLICK_EVENT_NAME)
+                        .setAllAttributes(createViewAttributes(view))
+                        .emit()
+                }
+            }
+        }
+    }
+
+    fun stopTracking() {
+        windowRef?.get()?.run {
+            if (callback is WindowCallbackWrapper) {
+                callback = (callback as WindowCallbackWrapper).unwrap()
+            }
+        }
+        windowRef = null
+    }
+
+    private fun createEvent(name: String): ExtendedLogRecordBuilder =
+        eventLogger
+            .logRecordBuilder()
+            .setEventName(name)
+
+    private fun createViewAttributes(view: View): Attributes {
+        val builder = Attributes.builder()
+        builder.put(viewNameAttr, viewToName(view))
+        builder.put(viewIdAttr, view.id.toLong())
+
+        builder.put(xCoordinateAttr, view.x.toDouble())
+        builder.put(yCoordinateAttr, view.y.toDouble())
+        return builder.build()
+    }
+
+    private fun viewToName(view: View): String =
+        try {
+            view.resources?.getResourceEntryName(view.id) ?: view.id.toString()
+        } catch (throwable: Throwable) {
+            view.id.toString()
+        }
+
+    private fun findTargetForTap(
+        decorView: View,
+        x: Float,
+        y: Float,
+    ): View? {
+        val queue = LinkedList<View>()
+        queue.addFirst(decorView)
+        var target: View? = null
+
+        while (queue.isNotEmpty()) {
+            val view = queue.removeFirst()
+            if (isJetpackComposeView(view)) {
+                return null
+            }
+
+            if (isValidClickTarget(view)) {
+                target = view
+            }
+
+            if (view is ViewGroup) {
+                handleViewGroup(view, x, y, queue)
+            }
+        }
+        return target
+    }
+
+    private fun isValidClickTarget(view: View): Boolean = view.isClickable && view.isVisible
```

Comment:
this can be a bit more clever. not sure if needed though.
eg https://github.com/PostHog/posthog-android/blob/d232586ce6b5043d716f108d06316d8de4714c31/posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt#L479-L520

Reply:
no need for that.